File: python/DSA/SingleLinkedList.py
class Node:
    def __init__(self, data):
        self.prev = None
        self.data = data
        self.next = None

class DoubleLinkedList:
    def __init__(self):
        self.head = None
        self.tail = None
    
    # to add value to beginning
    def add_begin(self, data):
        new_node = Node(data)
        if self.head is None:
            self.head = new_node
            self.tail = new_node
        else:
            self.head.prev = new_node
            new_node.next = self.head
            self.head = new_node
    
    # to add value in middle
    def add_middle(self, index, data):
        new_node = Node(data)
    
    # to add value to last
    def add_end(self, data):
        new_node = Node(data)
        self.tail.next = new_node
        new_node.prev = self.tail
        self.tail = new_node

    # to display the value
    def display(self, index = 0):
        if index >= 0:
            index_node = self.head
            while index_node is not None:
                print(index_node.data)
                index_node = index_node.next
        else:
            index_node = self.tail
            while index_node is not None:
                print(index_node.data)
                index_node = index_node.prev

    # to update the value 
    def update(self, data, new_data):
        index_node = self.head
        while index_node is not None:
            if index_node.data == data:
                index_node.data = new_data
            index_node = index_node.next
    
    # to remove the value by index
    def remove_index(self, index = 0):
        if index >= 0:
            if index == 0:
                self.head = self.head.next
            else:
                i = 0
                index_node = self.head
                while i < index and index_node.next is not None:
                    index_node = index_node.next
                    i+=1
                index_node.prev.next = index_node.next
        else:
            if index == -1:
                self.tail = self.tail.prev
                self.tail.next = None
            else:
                i = -1
                index_node = self.tail
                while i > index and index_node.prev is not None:
                    index_node = index_node.prev
                    i-=1
                index_node.next.prev = index_node.prev
                index_node.prev.next = index_node.next

    # to remove the value 
    def remove_value(self, value):
        index_node = self.head
        while index_node is not None:
            if index_node.data == value:
                if index_node.prev is not None:
                    index_node.prev.next = index_node.next
                else:
                    index_node.next.prev = None
                    self.head = index_node.next
                if index_node.next is not None:
                    index_node.next.prev = index_node.prev
            index_node = index_node.next

dd = DoubleLinkedList()
dd.add_begin(10)
dd.add_begin(20)
dd.add_begin(30)
dd.add_end(40)
dd.add_end(50)
dd.update(50, 60)
dd.remove_value(30)
dd.remove_index(-2)
dd.display(-1)

        

File: csharp/DSA/Algorithm/Collections/SingleLinkedList.cs
using BaseNamespace;

namespace Collections.Algorithms
{
	public class Node
	{
		public int Data { get; set; }
		public Node? Next { get; set; }
	}

	public static class SingleLinkedListExtension
	{
		public static void Sort(this SingleLinkedList singleLinkedList)
		{
			if (singleLinkedList != null)
			{
				Console.WriteLine("method called...");
				for (int Index = 0; Index < singleLinkedList.Count - 1; Index++)
				{
					for (int subIndex = 0; subIndex < singleLinkedList.Count - 1; subIndex++)
					{
						if (singleLinkedList[subIndex] > singleLinkedList[subIndex + 1])
						{
							(singleLinkedList[subIndex + 1], singleLinkedList[subIndex]) = (singleLinkedList[subIndex], singleLinkedList[subIndex + 1]);
						}
					}
				}
			}
		}
	}
	public class SingleLinkedList : IBluePrint
	{
		public Node? Header { get; set; }
		public int Count { get; set; }

		public SingleLinkedList()
		{
			Header = null;
		}

		// add to begin
		public void AddBegin()
		{
			Console.WriteLine("Enter the value: ");
			int data = Convert.ToInt32(Console.ReadLine());

			Node newNode = new()
			{
				Data = data,
				Next = Header
			};

			Header = newNode;
			Count += 1;
		}
		// add to middle
		public void AddMiddle()
		{
			Console.WriteLine("Enter the index: ");
			int index = Convert.ToInt32(Console.ReadLine());
			if (Header == null || index >= Count)
			{
				Console.WriteLine("Index Out of Range...");
			}
			else
			{
				Console.WriteLine("Enter the value: ");
				int data = Convert.ToInt32(Console.ReadLine());

				int currentIndex = 0;
				Node? currentNode = Header;

				while (currentIndex < index - 1)
				{
					currentNode = currentNode?.Next;
					currentIndex += 1;
				}
				if (currentNode != null)
				{
					Node newNode = new()
					{
						Data = data,
						Next = currentNode.Next
					};
					currentNode.Next = newNode;

					Count += 1;
				}
			}
		}
		// add to end
		public void AddEnd(int? Value)
		{
			int data;
			if (Value != null)
			{
				data = (int)Value;
			}
			else
			{
				Console.WriteLine("Enter the value: ");
				data = Convert.ToInt32(Console.ReadLine());
			}

			Node newNode = new()
			{
				Data = data,
				Next = null
			};

			if (Header == null)
			{
				Header = newNode;
			}
			else
			{
				Node currentNode = Header;

				while (currentNode.Next != null)
				{
					currentNode = currentNode.Next;
				}
				currentNode.Next = newNode;
			}
			Count += 1;
		}
		// display
		public void Display()
		{
			Node? currentNode = Header;

			while (currentNode != null)
			{
				Console.WriteLine(currentNode.Data);
				currentNode = currentNode.Next;
			}
		}
		// update by index
		public void UpdateIndex()
		{
			Console.WriteLine("Enter the Index: ");
			int index = Convert.ToInt32(Console.ReadLine());

			int NodeIndex = 0;
			Node? currentNode = Header;
			while (NodeIndex < index && currentNode != null)
			{
				currentNode = currentNode.Next;
				NodeIndex += 1;
			}

			if (currentNode != null)
			{
				Console.WriteLine("Enter the Value: ");
				currentNode.Data = Convert.ToInt32(Console.ReadLine());
			}
		}
		// update by value
		public void UpdateValue()
		{
			Console.WriteLine("Enter the Old Value: ");
			int OldValue = Convert.ToInt32(Console.ReadLine());

			Console.WriteLine("Enter the New Value: ");
			int NewValue = Convert.ToInt32(Console.ReadLine());

			Node? currentNode = Header;

			while (currentNode != null)
			{
				if (currentNode.Data == OldValue)
				{
					currentNode.Data = NewValue;
				}
				currentNode = currentNode.Next;
			}
		}
		// remove first
		public void RemoveFirst()
		{
			if (Header != null)
			{
				Header = Header.Next;
				Count -= 1;
			}
			else
			{
				Console.WriteLine("List is Empty...");
			}
		}

		// remove middle
		public void RemoveMiddle()
		{
			Console.WriteLine("Enter the Index: ");
			int index = Convert.ToInt32(Console.ReadLine());

			Node? currentNode = Header;
			int NodeIndex = 0;
			while (NodeIndex < index - 1)
			{
				currentNode = currentNode?.Next;
				NodeIndex += 1;
			}

			if (currentNode != null && currentNode.Next != null)
			{
				currentNode.Next = currentNode.Next.Next;
				Count -= 1;
			}
		}

		// remove last
		public void RemoveLast()
		{
			Node? currentNode = Header;

			while (currentNode?.Next?.Next != null)
			{
				currentNode = currentNode.Next;
			}

			if (currentNode != null)
			{
				currentNode.Next = null;
				Count -= 1;
			}
		}

		public void RemoveValue()
		{
			Node? currentNode = Header;

			Console.WriteLine("Enter the Value: ");
			int Value = Convert.ToInt32(Console.ReadLine());

			while (currentNode != null)
			{
				if (currentNode.Next != null && currentNode.Next.Data == Value)
				{
					currentNode.Next = currentNode.Next.Next;
				}
				currentNode = currentNode.Next;
			}
		}
		// to check value present or not
		public bool Contains(int? Data)
		{
			int Value;

			if (Data != null)
			{
				Value = (int)Data;
			}
			else
			{
				Console.WriteLine("Enter the value: ");
				Value = Convert.ToInt32(Console.ReadLine());
			}

			Node? currentNode = Header;
			while (currentNode != null)
			{
				if (currentNode.Data == Value)
				{
					return true;
				}
				currentNode = currentNode.Next;
			}

			return false;
		}

		// to find the value
		public int FindIndex()
		{
			Console.WriteLine("Enter the value: ");
			int Value = Convert.ToInt32(Console.ReadLine());
			int index = 0;

			Node? currentNode = Header;

			while (currentNode != null)
			{
				if (currentNode.Data == Value)
				{
					break;
				}
				index += 1;
				currentNode = currentNode.Next;
			}
			return index;
		}

		public int this[int Index]
		{
			get
			{
				Node? currentNode = Header;

				int NodeIndex = 0;
				while (currentNode != null && NodeIndex < Count)
				{
					if (Index == NodeIndex)
					{
						return currentNode.Data;
					}
					NodeIndex += 1;
					currentNode = currentNode.Next;
				}

				throw new IndexOutOfRangeException("Index Out of Range");
			}

			set
			{
				Node? currentNode = Header;

				int NodeIndex = 0;
				while (currentNode != null && NodeIndex < Count)
				{
					if (Index == NodeIndex)
					{
						currentNode.Data = value;
						return;
					}
					NodeIndex += 1;
					currentNode = currentNode.Next;
				}

				throw new IndexOutOfRangeException("Index Out of Range");
			}
		}

		public void AskQuestion()
		{
			SingleLinkedList linkedList = new();
			while (true)
			{
				Console.WriteLine("Enter the choice: ");
				string? choice = Console.ReadLine();
				if (choice == "addbeg")
				{
					linkedList.AddBegin();
				}
				else if (choice == "addmid")
				{
					linkedList.AddMiddle();
				}
				else if (choice == "addend")
				{
					linkedList.AddEnd(null);
				}
				else if (choice == "dis")
				{
					linkedList.Display();
				}
				else if (choice == "upin")
				{
					linkedList.UpdateIndex();
				}
				else if (choice == "upva")
				{
					linkedList.UpdateValue();
				}
				else if (choice == "remf")
				{
					linkedList.RemoveFirst();
				}
				else if (choice == "remm")
				{
					linkedList.RemoveMiddle();
				}
				else if (choice == "reml")
				{
					linkedList.RemoveLast();
				}
				else if (choice == "len")
				{
					Console.WriteLine($"Length: {linkedList.Count}");
				}
				else if (choice == "con")
				{
					if (linkedList.Contains(null))
					{
						Console.WriteLine("Value Found...");
					}
					else
					{
						Console.WriteLine("Value Not Found...");
					}
				}
				else if (choice == "find")
				{
					Console.WriteLine($"Index: {linkedList.FindIndex()}");
				}
				else if (choice == "getval")
				{
					Console.WriteLine("Enter the Index: ");
					Console.WriteLine(linkedList[Convert.ToInt32(Console.ReadLine())]);
				}
				else if (choice == "sort")
				{
					linkedList.Sort();
				}
				else if (choice == "exit")
				{
					break;
				}
				else
				{
					Console.WriteLine("Enter the correct choice");
				}
			}
		}
	}
}